android-components/components/support/remotesettings/src/main/java/mozilla/components/support/remotesettings/RemoteSettingsClient.kt

```diff
@@ -0,0 +1,109 @@
+import org.json.JSONObject
+import java.io.BufferedReader
+import java.io.File
+import java.io.FileReader
+import java.io.FileWriter
+import java.io.IOException
+
+/**
+ * Helper class to download collections from remote settings in app services.
+ *
+ * @property storageRootDirectory The top level app-local storage directory.
+ * @property serverUrl The optional url for the settings server. If not specified, the standard server will be used.
+ * @property bucketName The optional name of the bucket containing the collection on the server.
+ * If not specified, the standard bucket will be used.
+ * @property collectionName The name of the collection for the settings server.
+ */
+class RemoteSettingsClient(
+    private val storageRootDirectory: File,
+    private val serverUrl: String? = "https://firefox.settings.services.mozilla.com",
+    var bucketName: String? = "main",
+    var collectionName: String = "",
+) {
+
+    private val config = RemoteSettingsConfig(
+        serverUrl = serverUrl,
+        bucketName = bucketName,
+        collectionName = collectionName,
+    )
+
+    /**
+     * Fetches all records for a collection that can be found in the local storage.
+     **/
+    suspend fun fetch(): JSONObject? = withContext(Dispatchers.IO) {
+        val fileName = "list.json"
+        val localFile = File(storageRootDirectory, fileName)
+        if (localFile.exists()) {
+            try {
+                val fileReader = FileReader(localFile)
+                val bufferedReader = BufferedReader(fileReader)
+                val stringBuilder = StringBuilder()
+                var line: String?
+
+                while (bufferedReader.readLine().also { line = it } != null) {
+                    stringBuilder.append(line)
+                }
+
+                fileReader.close()
+
+                val jsonStr = stringBuilder.toString()
+                fetchRemoteServerRecords()
+                return@withContext JSONObject(jsonStr)
+            } catch (e: IOException) {
+                e.printStackTrace()
+                return@withContext null
+            } catch (e: JSONException) {
+                e.printStackTrace()
+                return@withContext null
+            }
+        } else {
+            return@withContext fetchRemoteServerRecords()
+        }
+    }
+
+    /**
+     * Fetches all records for the collection from remote servers and updates the local storage.
+     **/
+    suspend fun fetchRemoteServerRecords(): JSONObject = withContext(Dispatchers.IO) {
+        try {
+            val records = RemoteSettings(config).use {
+                it.getRecords().records.first().fields
+            }
+            updateLocalFile(records)
+            return@withContext records
+        } catch (e: RemoteSettingsException.ResponseException) {
+            JSONObject()
+        } catch (e: Exception) {
+            Log.d("exception",e.message.toString())
+            JSONObject()
+
+        }
+    }
```

Comment:
We should be careful about initiating disk writes without making that really clear what's happening in the API. We probably want to, at minimum, create separate `fetch` and `save` methods, though we could also consider a convenience function that did both.